### PR TITLE
Add Xerces-c

### PR DESCRIPTION
**Pull request by gillins**
Closes https://github.com/conda-forge/staged-recipes/pull/194

This PR adds our own version of Xerces-c that is built using `toolchain` fixing linking problems on OSX. 

I've added `icu` support which the default build didn't have.

ping @jakirkham

### CODE REVIEW

**Comment by ocefpaf** [comment]
@gillins do you want to remove the Windows support for now, get this merged, and then work on the feedstock?

**Comment by jakirkham** [comment]
Thanks for taking this over @gillins. 😄

**Comment by jakirkham** [comment]
@groutr, we are going to work on Xerces-C in this PR now.

**Comment by jakirkham** [comment]
Feel free to pilfer any useful changes from PR ( https://github.com/conda-forge/staged-recipes/pull/194 ), @gillins.

**Comment by gillins** [comment]
Sorry @groutr I didn't see your PR. Have pulled in your license section - let me know if there is anything else I am missing.

Good suggestion @ocefpaf I have disabled Windows and will look into it properly when I get a moment.

Have added @groutr, @ocefpaf and @jakirkham as maintainers. Please let me know if you are with this.

File: recipes/xerces-c/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+build:
+    number: 0
+    skip: True  # [win]
+
+requirements:
+    build:
+        - toolchain
+        - python  # [win]
+        - icu  # [not win]
```
Comment (jakirkham) [inline recipes/xerces-c/meta.yaml]: Please pin to `56.*`.

File: recipes/xerces-c/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+    run:
+        - icu  # [not win]
```
Comment (jakirkham) [inline recipes/xerces-c/meta.yaml]: Please pin to `56.*`.

File: recipes/xerces-c/build.sh [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+source activate "${CONDA_DEFAULT_ENV}"
+
+if [ "$(uname)" == "Darwin" ]
+then
+  export CXX="${CXX} -stdlib=libc++"
+fi
+
+./configure --prefix=$PREFIX --with-icu
+make
+make install
```
Comment (jakirkham) [inline recipes/xerces-c/build.sh]: Sorry, if I'm re-asking an old question, does this have some sort of `make check`, `make test`, or similar?

**Comment by jakirkham** [comment]
> Have added ... @jakirkham as maintainers. Please let me know if you are ok with this.

I'm not honestly that knowledgeable about this recipe nor have I used this package. Maybe I'm not a great maintainer. Sorry about that. 😞

**Comment by gillins** [comment]
Please don't merge - I just about have Windows support going....

**Comment by gillins** [comment]
@ocefpaf Can you restart AppVeyor? Thanks Seems to have hung just for Python34 x64. If this passes I think it is all good to merge.

**Comment by ocefpaf** [comment]
> @ocefpaf Can you restart AppVeyor? 

Done! I will check it again tomorrow. Thanks!

**Comment by gillins** [comment]
I'm getting the same error (below) and `%VS_MAJOR% isn't set causing things to go wrong. Have you seen this error in other builds? Just seems to be Python3.4 x64.

```
C:\Miniconda-x64\conda-bld\work\ianb-paste-768835f3c1b7>call "C:\Program Files (x86)\Microsoft Visual Studio 10.0\VC\vcvarsall.bat" amd64 
The specified configuration type is missing.  The tools for the
configuration might not be installed.
```

**Comment by jakirkham** [comment]
So, we do pin `conda-build` on Windows Python 3.4 32-bit. Maybe you could add something to the AppVeyor script to try updating `conda-build` after we pin it.

**Comment by gillins** [comment]
Since this is a staged recipe I don't think I have the power to change the AppVeyor pinning... Is that what you had in mind?

**Comment by ocefpaf** [comment]
Still hanging @gillins I will try one more time, but it seems to be a legit bug.

**Comment by gillins** [comment]
Same problem... Any suggestions?

**Comment by ocefpaf** [comment]
@gillins the failure is happening only for:

`Environment: TARGET_ARCH=x64, CONDA_PY=34, PY_CONDITION=python >=3.4,<3.5`

and we have this in `conda-forge-build-setup`:

```shell
:: Workaround for Python 3.4 and x64 bug in latest conda-build.
:: FIXME: Remove once there is a release that fixes the upstream issue
:: ( https://github.com/conda/conda-build/issues/895 ).
if "%TARGET_ARCH%" == "x64" if "%CONDA_PY%" == "34" conda install -n root --yes --quiet conda-build=1.20.0
```

To test is this version is the problem I recommend updating conda-build to the latest version right after line https://github.com/conda-forge/staged-recipes/blob/master/appveyor.yml#L77 and, if that works, fixing `conda-forge-build-setup`.

**Comment by gillins** [comment]
Thanks @ocefpaf - fingers crossed!

**Comment by ocefpaf** [comment]
It does not look promising :unamused:

It seems to be already stuck at the same point:

```shell

C:\Miniconda-x64\conda-bld\work\xerces-c-3.1.3>if not exist projects\Win32\VC\xerces-all\xerces-all.sln (
REM see https://github.com/conda/conda-build/wiki/Windows-recipe-patterns#updating-older-sln-files  
 REM Upgrade the VC9 version - got weird errors with VC12  
 REM I should be able to update XERCESVC above, but it doesn't work for me !?  
 devenv.exe projects\Win32\VC9\xerces-all\xerces-all.sln /Upgrade  
 if errorlevel 1 exit 1  
 msbuild projects\Win32\VC9\xerces-all\xerces-all.sln /t:XercesLib /p:Configuration=Release;Platform=x64  
 if errorlevel 1 exit 1  
 REM implib and DLL - this method puts it in a different place?!  
 copy projects\Win32\VC9\xerces-all\XercesLib\Win64\xerces-c_3.lib C:\Miniconda-x64\envs\_build\Library\lib  
 if errorlevel 1 exit 1  
 copy projects\Win32\VC9\xerces-all\XercesLib\Win64\xerces-c_3_1.dll C:\Miniconda-x64\envs\_build\Library\bin  
 if errorlevel 1 exit 1 
)  else (
msbuild projects\Win32\VC\xerces-all\xerces-all.sln /t:XercesLib /p:Configuration=Release;Platform=x64  
 if errorlevel 1 exit 1  
 REM implib and DLL  
 copy Build\Win64\VC\Release\xerces-c_3.lib C:\Miniconda-x64\envs\_build\Library\lib  
 if errorlevel 1 exit 1  
 copy Build\Win64\VC\Release\xerces-c_3_1.dll C:\Miniconda-x64\envs\_build\Library\bin  
 if errorlevel 1 exit 1 
) 
```

@patricksnape and @msarahan any ideas?

@gillins maybe we should just skip this item of the matrix get `xerces` into the feedstock.

**Comment by gillins** [comment]
It looks like an AppVeyor problem to me:
```
C:\Miniconda-x64\conda-bld\work\xerces-c-3.1.3>call "C:\Program Files (x86)\Microsoft Visual Studio 10.0\VC\vcvarsall.bat" amd64 
The specified configuration type is missing.  The tools for the
configuration might not be installed.
```

**Comment by ocefpaf** [comment]
> It looks like an AppVeyor problem to me

Argh... Not sure how to proceed beyond skipping that build for now.

File: appveyor.yml [hunk 4]
```diff
@@ -75,6 +75,7 @@ install:
 
     - cmd: conda install --yes --quiet obvious-ci conda-build-all
     - cmd: conda install --yes --quiet conda-forge-build-setup
+    - cmd: conda update conda-build
```
Comment (ocefpaf) [inline appveyor.yml]: Let's remove this too.

**Comment by msarahan** [comment]
@gillins I'm not sure you can trust that message. Our strategy with activating compiler is to try all the scripts possible.  You may be seeing a message that means it is trying the equivalent of vs 2010 pro and not finding it. 

Have a look at the msvc tests in conda build:

https://github.com/conda/conda-build/blob/1.21.x/tests/test_build_recipes.py#L208

https://github.com/conda/conda-build/blob/1.21.x/tests/test-recipes/metadata/_build_msvc_compiler/bld.bat

If you have ideas for improving things, let me know. One possibly important difference with the tests is that they are always overriding distutils logic by setting MSSdk and DISTUTILS_USE_SDK.

**Comment by msarahan** [comment]
@gillins don't work too hard on this yet.  There are a couple of issues that may be at play here:

 * https://github.com/conda/conda/pull/2917 - if an activation step comes before the MSVC activate scripts, then we can lose the activation path.
 * https://github.com/conda/conda-build/issues/1048 - I don't think this one is affecting us on AppVeyor, but there were issues with Py3 builds due to compiler finding logic.

Of course, neither of these account for hangs, so maybe you do want to dig in after all?

**Comment by gillins** [comment]
@msarahan I was hoping to show you the behaviour I was seeing. None of the variables (`%VS_MAJOR%` etc) were being set causing `bld.bat` to use invalid paths. So something wasn't quite right in the build process.

I think the hang was maybe due to a message box being shown, but I'm just guessing.

**Comment by msarahan** [comment]
That's cool too.  I'm happy to take a look when AppVeyor finishes - perhaps
tomorrow.

**Comment by gillins** [comment]
@msarahan see https://ci.appveyor.com/project/conda-forge/staged-recipes/build/1.0.3887/job/2wnhd0e5qcl866cb#L812. The `bld.bat` line is `set XERCESVC=VC%VS_MAJOR%` and should be `VC10`....

**Comment by msarahan** [comment]
Yeah, that's wrong, but I think that's due to pinning an old conda-build version: https://ci.appveyor.com/project/conda-forge/staged-recipes/build/1.0.3887/job/2wnhd0e5qcl866cb#L212

Those definitions only came around with conda-build 1.20.1 : https://github.com/conda/conda-build/blob/1.20.0/conda_build/windows.py
https://github.com/conda/conda-build/blob/1.20.1/conda_build/windows.py

Probably just a red herring, unfortunately.  Try unpinning conda-build again, and see if we get anything different.

**Comment by msarahan** [comment]
PS: your earlier effort to unpin (https://github.com/conda-forge/staged-recipes/pull/849/commits/d7711c1f762ce92de28e2cee761ae9a807306dde) was unsuccessful: things got downgraded due to dependencies somehow: https://ci.appveyor.com/project/conda-forge/staged-recipes/build/1.0.3866/job/3s46cxjfsgbpoxbn#L169

**Comment by gillins** [comment]
Hmmm. Same thing.... Weird that Python 3.4 32bit seems to get the right version.

File: appveyor.yml [hunk 5]
```diff
@@ -75,6 +75,7 @@ install:
 
     - cmd: conda install --yes --quiet obvious-ci conda-build-all
     - cmd: conda install --yes --quiet conda-forge-build-setup
+    - cmd: conda update -n root conda-build
     - cmd: run_conda_forge_build_setup
```
Comment (jakirkham) [inline appveyor.yml]: We need to add the update after running this script otherwise it won't work.

**Comment by jakirkham** [comment]
I canceled the old AppVeyor build as it was hanging, near max time, and the new commit had been sent. Let's see if this does better. If not, please let me know.

File: appveyor.yml [hunk 6]
```diff
@@ -75,8 +75,8 @@ install:
 
     - cmd: conda install --yes --quiet obvious-ci conda-build-all
     - cmd: conda install --yes --quiet conda-forge-build-setup
-    - cmd: conda update -n root conda-build
     - cmd: run_conda_forge_build_setup
+    - cmd: conda update -n root conda-build
```
Comment (jakirkham) [inline appveyor.yml]: 👍

File: appveyor.yml [hunk 7]
```diff
@@ -76,6 +76,7 @@ install:
     - cmd: conda install --yes --quiet obvious-ci conda-build-all
     - cmd: conda install --yes --quiet conda-forge-build-setup
     - cmd: run_conda_forge_build_setup
+    - cmd: conda update -n root conda-build
```
Comment (jakirkham) [inline appveyor.yml]: We may also need to add `conda-build-all` here. This needs `--yes` too.

**Comment by jakirkham** [comment]
Had to cancel AppVeyor as the jobs were waiting for confirmation. We need the `--yes` flag.

File: appveyor.yml [hunk 8]
```diff
@@ -76,7 +76,8 @@ install:
     - cmd: conda install --yes --quiet obvious-ci conda-build-all
     - cmd: conda install --yes --quiet conda-forge-build-setup
     - cmd: run_conda_forge_build_setup
-    - cmd: conda update -n root conda-build
+    - cmd: conda update --yes -n root conda-build
+    - cmd: conda-build-all --yes
```
Comment (jakirkham) [inline appveyor.yml]: It should be `conda update --yes -n root conda-build conda-build-all`.

**Comment by jakirkham** [comment]
That seems to be upgrading correctly. However, it looks there are some build issues that are coming out now.

**Comment by gillins** [comment]
Seems that the latest version of `xerces-c` has a broken project file. I'll investigate. Not sure about the other error though.
```
projects\Win32\VC9\xerces-all\xerces-all.sln(1): Solution file error MSB5014: File format version is not recognized.  
MSBuild can only read solution files between versions 7.0 and 9.0, inclusive.
```

**Comment by jakirkham** [comment]
So, Windows Python 3.4 32-bit failed, but not for any real reason. Instead it failed due to a 404 error encountered when trying to download a package. I went ahead and restarted the build. Unfortunately, that means restarting the full matrix. 😒

I might need to call it soon. However, it seems the pinning issues have been worked around successfully, which was the main concern. Will check back in the morning to see how things are going. Best of luck in the interim.

**Comment by jakirkham** [comment]
FYI sounds like we need a newer `msbuild`. ( http://stackoverflow.com/a/8177500 ) We can actually build it ourselves. ( https://github.com/Microsoft/msbuild ) However, it might be best to just skip that build for now and save it for the feedstock if that seems reasonable.

**Comment by gillins** [comment]
I don't get what is going on there since the 64 bit version works, but the 32 bit fails. They are both using the same version of VC in theory.

**Comment by jakirkham** [comment]
> I don't get what is going on there since the 64 bit version works, but the 32 bit fails. They are both using the same version of VC in theory.

Hmm...pretty sure there was some funky stuff going on to get VC 2008 64-bit to work. In particular, I think it uses MVC++ compiler for Python 2.7 to get the 64-bit support. Whereas 32-bit is going through a different path. Though I don't really remember all the details nor am I an expert on this. Just trying to recount the history that is coming to me.

In any event, everything else passed this time. So, the unpinning works. Particularly for addressing Python 3.4 32-bit.

**Comment by gillins** [comment]
Thanks everyone for helping with this. Seems we are very close - what should I try for getting Python 2.7 32bit going? Has anyone come up against this before?

**Comment by msarahan** [comment]
See https://github.com/conda-forge/staged-recipes/pull/194#issuecomment-203577297

**Comment by gillins** [comment]
Hmmm.... So is this something I can workaround in `bld.bat` or does it need to be a change in `conda-build`?

**Comment by msarahan** [comment]
Let's play in bld.bat for now. If it works, will add it to conda build.

**Comment by gillins** [comment]
So do I have to set that `Framework35Version` variable, or change the whole path?

**Comment by msarahan** [comment]
I don't know, and his comment doesn't make it completely clear.  Try just setting the variable first.  Failing that, we'll have to prepend PATH with some other known good location for a newer msbuild.  That's going to probably be something only on the conda-forge side, since it is too hard to do it for arbitrary build environments.

**Comment by msarahan** [comment]
Oh dear, it looks like this PR picked up other recipes from master, perhaps?  I don't see any reason why snappy should be a dependency of this recipe.

**Comment by ocefpaf** [comment]
Travis-CI is playing a number on us :unamused: 

@pelson is there is git/github trick that we could do to avoid committing recipes to master? Just so we always have a pristine branch. Not sure if that is possible but something along the lines:

- master has only the examples and the CI boilerplate;
- people start their PRs from `master` but the PRs should be directed to a"feedstocking" branch instead;
- everything that happens in master now would happen in that branch "feedstocking" branch.

Not sure if that is a good idea though... If that is possible it will add new step to point the PRs to an unusual branch (not `master`) therefore making new contributions harder than it should be.

**Comment by jakirkham** [comment]
I was having the same thought some weeks back. We definitely should do this. Might try to look at something later tonight, but if I can't get it done. I'll try to look at it when I get back probably on Wednesday (if someone doesn't beat me to it).

**Comment by jakirkham** [comment]
Could you please put that idea in a new issue so we don't lose it in this thread, @ocefpaf? 😄

**Comment by pelson** [comment]
@gillins - my best advice for moving this forwards would be to bank what you have for windows, but skip it in the configuration. One motivation is to help us move on with the libstdc++ issue. It is interesting to note that the `defaults` version will be linking to `libstdc++` while the conda-forge one would not - I'm uncertain what impact that could have.

**Comment by jakirkham** [comment]
I added your idea to this issue ( https://github.com/conda-forge/staged-recipes/issues/942 ) @ocefpaf.

**Comment by ocefpaf** [comment]
> Could you please put that idea in a new issue so we don't lose it in this thread, @ocefpaf? 😄

It seems you already got to it.

> @gillins - my best advice for moving this forwards would be to bank what you have for windows, but skip it in the configuration.

:+1: 

> One motivation is to help us move on with the libstdc++ issue. It is interesting to note that the `defaults` version will be linking to `libstdc++` while the conda-forge one would not - I'm uncertain what impact that could have.

I am afraid of that impact but it is worth a try.

@msarahan what are your thoughts on this? Any special reason why `defaults` went for `libstdc++`? Or nothing special?

**Comment by msarahan** [comment]
> msarahan what are your thoughts on this? Any special reason why defaults went for libstdc++? Or nothing special?

Not that I know of. Perhaps just lack of knowledge of attention to detail. 

IMHO, with channel priority and with consistent stacks on either side, the should be no issues.

**Comment by jakirkham** [comment]
>> msarahan what are your thoughts on this? Any special reason why defaults went for libstdc++? Or nothing special?

> Not that I know of. Perhaps just lack of knowledge or attention to detail.

If I had to guess, Continuum was supporting 10.6 before and as such had to use `gcc`/`libstdc++` (there was no `clang`/`libc++` for 10.6). Once they switched to 10.7, keeping `gcc`/`libstdc++` would have been sensible to avoid breakage. Though IMHO there would not be anything blocking a switch to `clang` if that were desired.

File: appveyor.yml [hunk 9]
```diff
@@ -76,6 +76,7 @@ install:
     - cmd: conda install --yes --quiet obvious-ci conda-build-all
     - cmd: conda install --yes --quiet conda-forge-build-setup
     - cmd: run_conda_forge_build_setup
+    - cmd: conda update --yes -n root conda-build conda-build-all
```
Comment (jakirkham) [inline appveyor.yml]: Probably should remove this now that we are done experimenting.

**Comment by jakirkham** [comment]
Is this still WIP, @gillins?

With one minor note about the AppVeyor CI script, I would be more than happy to see this merged.

It looks like it is doing the right stuff w.r.t. `libc++`. Even if we find other issues, it will be easier to address once we have a binary to play with.

Thoughts?

**Comment by gillins** [comment]
Yes I think it is ready to merge. I'll work on the Windows support once it is a feedstock.

**Comment by ocefpaf** [comment]
Thanks @gillins! I know this one was painful!! (And it is not over yet...)

**Pull request merged**